pkg/controller/util.go

```diff
@@ -109,8 +110,158 @@ func readDecodedRole(s string) (*spec.PgUser, error) {
 	return &result, nil
 }
 
-func (c *Controller) getInfrastructureRoles(rolesSecret *spec.NamespacedName) (map[string]spec.PgUser, error) {
-	if *rolesSecret == (spec.NamespacedName{}) {
+var emptyName = (spec.NamespacedName{})
+
+// Return information about what secrets we need to use to create
+// infrastructure roles and in which format are they. This is done in
+// compatible way, so that the previous logic is not changed, and handles both
+// configuration in ConfigMap & CRD.
+func (c *Controller) getInfrastructureRoleDefinitions() []*config.InfrastructureRole {
+	var roleDef config.InfrastructureRole
+	rolesDefs := c.opConfig.InfrastructureRoles
+
+	if c.opConfig.InfrastructureRolesSecretName == emptyName {
+		// All the other possibilities require secret name to be present, so if
+		// it is not, then nothing else to be done here.
+		return rolesDefs
+	}
+
+	// check if we can extract something from the configmap config option
+	if c.opConfig.InfrastructureRolesDefs != "" {
+		// The configmap option could contain either a role description (in the
+		// form key1: value1, key2: value2), which has to be used together with
+		// an old secret name.
+
+		propertySep := ","
+		valueSep := ":"
+
+		// The field contains the format in which secret is written, let's
+		// convert it to a proper definition
+		properties := strings.Split(c.opConfig.InfrastructureRolesDefs, propertySep)
+
+		roleDef = config.InfrastructureRole{
+			SecretName: c.opConfig.InfrastructureRolesSecretName,
+			Template:   false,
+		}
+
+		for _, property := range properties {
+			values := strings.Split(property, valueSep)
+			if len(values) < 2 {
+				continue
+			}
+			name := strings.TrimSpace(values[0])
+			value := strings.TrimSpace(values[1])
+
+			switch name {
```

Comment:
if you would also extract the `secretname` key here, we could mark the `infrastructure_roles_secret_name` as deprecated and one could solely rely on `infrastructure_roles_secrets`. The surrounding conditions need to be adapted then:

```
if c.opConfig.InfrastructureRolesSecretName == emptyName {
  if c.opConfig.InfrastructureRolesDefs == "" {
    return rolesDefs
  } else {
    do extraction from configmap

} else {
  older format
} 
```

Now, it's a bit confusing that you have to use both config options together when dealing with configmap. But if you do the same for CRD only the old format is used and the new format is ignored (see `if c.opConfig.InfrastructureRolesSecretName == emptyName`)